### PR TITLE
feat(dagster-k8s): inject dagster session env vars to all containers

### PR DESCRIPTION
**Pull request by aignas**
## Summary & Motivation

This PR adds a way for users to be able to read to the pipes session context from containers 
other than the first container in the pod spec allowing sidecars to read extra information from Dagster. This PR also allows to make it possible to write messages to a file if needed,
but we default to writing to `/dev/null` to keep `open_dagster_pipes` functional without 
any extra need from the user.

The user can specify a custom message writer by specifying it like below:
```python
from dagster_pipes import encode_env_var

# Set the env like this in the `base_pod_spec`
    env = [
        {
            "name": "DAGSTER_PIPES_MESSAGES",
            "value": encode_env_var({"path": "/custom/path"}),
        },
    ],
```

## How I Tested These Changes

I have added unit test coverage to the necessary places and whilst doing it I have split
the tests to provide better feedback to the developer debugging the failures. I have tested
the patch internally to see that it is functional with our dagster pipes setup.

### CODE REVIEW

File: python_modules/libraries/dagster-k8s/dagster_k8s/pipes.py [hunk 1]
```diff
@@ -303,24 +305,54 @@ def build_pod_body(
...
+            raise DagsterInvariantViolationError(
+                "Should specify 'command' via 'base_pod_spec' when specifying multiple containers"
+            )
 
-    if image:
-        spec["containers"][0]["image"] = image
+        for container in (containers + init_containers):
+            for key in ["name", "image"]:
+                if key not in container:
+                    raise DagsterInvariantViolationError(f"Must provide base_pod_spec with {key} property set.")
```
Comment (alangenfeld) [inline python_modules/libraries/dagster-k8s/dagster_k8s/pipes.py:341]: not critical: would be nice if the error message pointed you at which container / init_container had the problem, maybe just by index?

File: python_modules/libraries/dagster-k8s/dagster_k8s/pipes.py [hunk 2]
```diff
@@ -303,24 +305,54 @@ def build_pod_body(
...
 
-    if "env" not in spec["containers"][0]:
-        spec["containers"][0]["env"] = []
+    # Substitute the Dagster session variables in the pod spec
+    injected_pipes_session_vars = False
+    for container in (containers + init_containers):
+        for env in container.get("env", []):
+            if env["name"] in [DAGSTER_PIPES_CONTEXT_ENV_VAR, DAGSTER_PIPES_MESSAGES_ENV_VAR]:
+                injected_pipes_session_vars = True
+                env["value"] = env_vars[env["name"]]
```
Comment (alangenfeld) [inline python_modules/libraries/dagster-k8s/dagster_k8s/pipes.py:352]: should probably call this out in a doc block somewhere, maybe as a note in the `base_pod_spec` arg on `run`?

**Review by alangenfeld — COMMENTED** [review]
Interesting PR! I can see how injecting the context in to multiple containers could be useful but I think trying to run the message writer from two places would lead to them overwriting each other or just being silently ignored in the non-"primary" containers. 

Did you run something with this set-upt? Curious to hear about what you tried and what worked if so. 

`make ruff` in repo root should fix the ruff step that failed, picking up latest changes from `master` will fix the docs

**Comment by aignas** [comment]
Hi @alangenfeld, thanks for the review. At work we are using Dagster to spin up pods in kubernetes where we have an init container that processes input set via the context and does some initial setup, then the main container runs (which does not know about Dagster) and then there is a sidecar container that does work after the main container exits and reports back to Dagster.

Note, the init container has to connect to the pipes session without closing when exiting for it to work and if the sidecar container connects to Dagster before the last message from the init container is received, we get warnings in Dagster UI about ignored messages. Everything is working (we have a few other patches to the dagster kubernetes client to make everything more reliable, they will be submitted later).

You mentioning the message writers overwriting each other is what we see in practice and I wonder if this PR should also change the message writers doing multiplexing of sorts where Dagster can receive messages through all connections and send messages by selecting the first container that has a live connection or selecting the container by its name in the pod spec. I'd be happy to do that in a followup PR if you could give me some pointers on where to start if this technical solutions sounds like the right direction to investigate.

P.S. I'm in JST timezone, so I'll address review comments a little later during the day.

**Comment by alangenfeld** [comment]
Which message reader/writer are you using for context? 

I suspect that the message reader / writer changes to make this work in general would be quite involved and considering the breadth of integrations we aspire to cover with pipes would be particularly burdensome complexity. 

I wonder if we can get to a setup where we can inject the real context and real messages config in to a single "primary container" and then provide the real context and a fake/pipe-to-dev-null messages config to any number of "auxiliary containers". It sounds like in your set-up its only that sidecar container that really needs to communicate back to Dagster.

**Comment by aignas** [comment]
Makes sense, we can get away with having just one of the containers communicate back with the control plane.

So the idea would be to implement my own `MessageWriter` that I can use to initiate the pipes session from the pod? I would still need to get the context environment variable templating done similarly how this PR is doing it. I could leave only `DAGSTER_PIPES_CONTEXT` templating in this PR if that is the way forward you would like to take.

I'll test this today and see if that works for our usecase.

**Comment by aignas** [comment]
@alangenfeld, thanks for the comments, I have updated the PR to not do replacement, but to instead work like you have suggested. The design seems to be much better right now! Let me know if you have anything else that you would like to see changed.

**Review by alangenfeld — APPROVED** [review]
nice, glad this worked I think this is pretty slick

**Pull request merged**